### ISSUE TITLE
深度学习模型与大模型

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#75]
# 理解模型架构

模型架构是指在机器学习和深度学习中，定义模型结构和组织方式的框架，它包括层的类型、排列顺序、超参数、输入和输出格式、训练机制等：

1. 层的类型：如卷积层、循环层、全连接层等，每种负责特定的功能：
    1. 卷积层（Convolutional Layer）：用于提取输入数据中的局部特征，特别适用于图像处理。工作原理：通过卷积操作使用多个滤波器（或核）扫描输入数据，生成特征图。这些滤波器能够捕捉不同的特征，如边缘、纹理等。
    2. 循环层（Recurrent Layer）：用于处理序列数据，如时间序列和自然语言，能够捕捉时间上的依赖关系。工作原理：通过在时间步之间传递隐藏状态，循环神经网络（RNN）能够记住之前的输入信息，从而有效处理上下文信息。常见的变种有 LSTM（长短时记忆网络）和 GRU（门控循环单元），它们在处理长序列时表现更好。
    3. 全连接层（Fully Connected Layer）：用于将提取的特征映射到最终的输出类别或回归值。工作原理：每个神经元与前一层的所有神经元相连接，通常位于网络的最后几层。通过加权求和和激活函数，将输入特征转换为输出。

2. 层的排列顺序：层与层之间的连接方式，例如，如何将一个层的输出传递到下一个层。
3. 超参数：如层的数量、每层的神经元数量、学习率、激活函数、词汇表大小等，这些参数影响模型的性能和训练过程。
4. 输入和输出格式：定义模型接受什么形式的输入，以及输出结果的形式
5. 训练机制：如何通过反向传播和优化算法来调整模型参数，以提高性能。

模型架构的设计直接影响模型的学习能力、泛化能力和计算效率。不同的任务和数据集通常需要不同的架构来达到最佳效果。

常见的模型架构有：BERT、GPT、Transformer-XL、RoBERTa、ALBERT、XLNet、T5、ERNIE、DistilBERT。阿里开源的 Qwen 2.5 系列的架构是 `Qwen2ForCausalLM`，我们以 Qwen2.5-7b-instruct 模型来说明模型架构。

## Qwen2ForCausalLM 架构

```
  | "architectures": [
  | "Qwen2ForCausalLM"
  | ],
  | "attention_dropout": 0.0,
  | "bos_token_id": 151643,
  | "eos_token_id": 151645,
  | "hidden_act": "silu",
  | "hidden_size": 3584,
  | "initializer_range": 0.02,
  | "intermediate_size": 18944,
  | "max_position_embeddings": 32768,
  | "max_window_layers": 28,
  | "model_type": "qwen2",
  | "num_attention_heads": 28,
  | "num_hidden_layers": 28,
  | "num_key_value_heads": 4,
  | "rms_norm_eps": 1e-06,
  | "rope_theta": 1000000.0,
  | "sliding_window": 131072,
  | "tie_word_embeddings": false,
  | "torch_dtype": "bfloat16",
  | "transformers_version": "4.43.1",
  | "use_cache": true,
  | "use_sliding_window": false,
  | "vocab_size": 152064
  | }
```

基本配置：
- CausalLM: Casual Language Modeling（因果语言建模）
- model_type: 模型的类型
- transformers_version: 依赖 Transformer 库的版本

模型参数：
- hidden_size: 隐层向量的维度，值为 3584
- num_hidden_layers: 隐层的数量，即Transformer 的层数，这里是 28层
- num_attention_heads: 注意力机制中的头数，每层 Transformer 的自注意力机制包含 28 个注意力头
- num_key_value_heads: key和value的头数，表示在多头注意力机制中Q、K、V拆分为较少的头数，从而优化性能
- intermediate_size: 前馈神经网络层（FFN）的中间层维度，值为18944.通常比hidden_size 大，提供更强的表达能力。FFN是每个Transformer层的重要组成部分，起着增强特征、提高模型容量的作用，是Transformer 模型提取语义信息的关键部分之一

数据类型：
- torch_dtype, 指定模型权重和计算的精度，这里是 bfloat16

缓存：
- use_cache: 是否在生成时缓存 key/value，以便于加速推理

词汇表：
- vocab_size: 模型可以处理的唯一词汇数量。较大的词汇表能让模型处理更多不同的单词、符号等，提高对长尾词汇或专有词汇的处理能力。但模型的embedding layer和输出投影矩阵的参数量也会更多，增加了模型的复杂度和内存需求。

位置编码与窗口：
- max_position_embeddings: 模型可以处理的最大位置编码数，即输入的最大序列长度是 32768
- max_window_layers: 支持滑动窗口的Transformer 层数。它是一种在处理超长序列时优化计算和内存使用的技术。它通过分块处理序列并限制每个位置的注意力范围，从而减少计算复杂度和内存需求，同时保留模型的上下文理解能力。

模型正则化：
- attention_dropout: 在注意力机制中使用的 Dropout 概率，用于防止过拟合。这里的 0.0 表示没有 Dropout。

## 为什么LLM 有max tokens的限制？

doing

## Mixture of Experts(MoE)

![Image](https://github.com/user-attachments/assets/55b03095-d9c3-4d7e-b61b-f1d2050d5c21) [^1]

MoE 的核心思想是将多个专家网络（子模型）组合在一起，通过一个门控机制（Gating Network）动态选择部分专家进行计算，而不是所有专家都参与推理和训练。MoE 至少有三个优势：
1. 低计算成本：每次激活的仅是部分网络，计算成本显著降低；
2. 高可扩展性：可以扩展到万亿参数，但计算时只有部分参与计算；
3. 任务专业化：每次任务激活的是专业的子模型，更为准确，不受噪音干扰。

**门控网络**

门控网络是一个比较重要的组件，它的路由准确度直接影响效果：通常是一个小型神经网络（如Softmax或top-k路由），生成每个专家的权重，决定哪些专家会被激活。

1. Softmax 实现（TensorFlow）

```
import tensorflow as tf

def softmax_gating(inputs, num_experts):
    """softmax gating function for routing inputs to experts."""
    gating_logits = tf.keras.layers.Dense(num_experts)(inputs)
    gating_weights = tf.nn.softmax(gating_logits)
    return gating_weights
```

实现方式：简单的 Dense 层 + Softmax，计算每个专家的权重。
特点：
- 适用于 Dense MoE（所有专家都有非零权重）
- 计算量较大，不适用于超大规模模型。

2. Top-K 实现（DeepSpeed）

```
def topk_gating(logits, k=2):
    """top-k gating for sparse activation in MoE."""
    top_k_logits, top_k_indices = torch.topk(logits, k, dim=-1)
    top_k_gates = torch.nn.functional.softmax(top_k_logits, dim=-1)
    return top_k_gates, top_k_indices
```

实现方式：
- torch.topk() 选择得分最高的k个专家
- 仅对这k个专家计算Softmax归一化
特点：
- 计算量小，适用于超大规模MoE
- k=2代表每次仅激活2个专家（可调节）。

# 深度学习模型与大模型的区别与联系

大模型是深度学习模型的特定进化形式，它们都基于深度学习的技术原理，但在规模、应用场景和计算需求上有所不同。

联系：

- 基于相同的技术原理：深度学习模型和大模型都基于人工神经网络，特别是多层网络的架构。它们使用相似的训练方法，如梯度下降、反向传播等，并依赖于大规模的数据来学习复杂的特征和模式。
- 训练方式：无论是深度学习模型还是大模型，都依赖于监督学习、无监督学习、强化学习等常见的训练方式。训练过程通常通常需要大量的数据和强大的计算资源。
- 框架与工具：它们通常使用相同的深度学习框架来构建和训练，例如TensorFlow、PyTorch等。大规模的构建和训练过程仍然依赖于这些框架，只是规模和复杂度更大。
- 应用领域：都广泛应用于人工智能的多个领域，包括计算机视觉、自然语言处理、语音识别、推荐系统等。大模型的出现并没有改变应用的根本领域，而是提升了模型在这些领域的性能和应用范围。

区别：

1. 规模：深度学习模型可以具有不同规模的网络架构，从相对简单的多层感知机（MLP）到复杂的卷积神经网络（CNN）、循环神经网络（RNN）等。这些模型的参数量从几万到几亿不等，适用于较小规模的任务和应用；大模型则是深度学习模型中的一种特例，其特点是拥有极其庞大的参数量。例如 ，OpenAI的GPT-3拥有1750亿参数，而GPT-4甚至可能达到万亿参数。大模型的训练需要及其庞大的计算资源和数据

2. 复杂度：传统的深度学习模型结构相对较为简单，通常针对特定任务进行优化和设计。例如CNN用于图像处理，RNN用于时间序列或序列数据处理。虽然它们可以解决复杂问题，但大多专注于单一任务；大模型通常基于Transformer等高度复杂的网络架构，具备更广泛的能力和应用场景。它们能处理更复杂的任务，并且往往具有多任务学习和生成能力。例如，GPT-3和GPT-4能够进行文本生成、语言翻译、编写代码等多种任务，而不局限于某一个领域

3. 数据与训练资源：传统的深度学习模型可以在相对较少的数据集和有限的计算资源下进行训练。例如，传统GPU或CPU足以训练一些中小规模的模型；大模型需要大量的数据来进行训练，通常是大规模的预料库或其他领域的海量数据。训练这些模型需要分布式计算、专用硬件（如TPUs、超算集群）以及长时间的计算资源投入，通常只能由大公司或研究机构完成，甚至研究机构也很难进行研究，参考李飞飞的采访

4. 通用性：传统深度学习模型通常是为特定任务设计的，适应性相对有限。一个图像分类模型不能直接用于自然语言处理，模型需要为不同任务进行重新设计和训练；大模型具有很强的通用性，尤其是在自然语言处理（NLP）领域。比如GPT-3或GPT-4可以用于文本生成、问答、翻译、对话等多种任务。这类大模型可以在没有专门训练的情况下，通过微调（FineTune）或上下文信息（组织Prompt）处理多样化的任务，具备通用智能的潜力

5. 部署与应用：传统深度学习模型的部署相对简单，可以嵌入到各种设备中，如移动设备、嵌入式系统等。应用程序的部署和推理速度也较为快速和轻量；大模型由于其规模和复杂度，通常难以直接部署到资源有限的设备上。它们更多地部署在云端服务中，提供API以供用户调用。大模型的推理速度也较慢，通常需要更高的计算能力，尤其是在实时性要求较高的场景下

todo：分布式部署技术解决大模型部署？
todo：量化&剪枝解决性能不够问题？

6. 研究前沿：深度学习模型的研究仍然非常活跃，尤其是在优化训练过程、提高模型可解释性、减少训练时间等方面。它们仍然是解决很多实际问题的主流方法；大模型处于当前深度学习研究的前沿，尤其是在NLP、生成模型（如生成对抗网络，GAN）、自动编程等领域。它们推动了对通用人工智能（AGI）的探索，并且大模型的应用潜力和挑战成为了现代AI研究的焦点之一

## 大模型的应用领域

1. 自然语言处理（NLP）：大模型在 NLP 中展现了极高的潜力，能够理解、生成和翻译多种语言文本。它们的广泛应用包括：

    1. 文本生成：可以自动撰写文章、博客、新闻报道、技术文档等。GPT 系列大模型的生成能力非常强大，能够根据给定的提示生成连贯且高质量的文本。
    2. 语言翻译：大模型可以进行跨语言翻译，提供高精度的机器翻译服务，甚至可以理解复杂的上下文和语境。
    3. 对话系统：大模型可以用于构建智能的聊天机器人和虚拟助手，如客服系统、医疗咨询、教育问答等。它们能够进行多轮对话，并且理解上下文语义。
    4. 情感分析和文本分类：大模型能够进行情感分析、主题分类、内容推荐等复杂的文本处理任务。

2. 自动编程与代码生成：大模型在代码生成和编程领域展现出了巨大的潜力，特别是对于开发者的支持：

    1. 代码生成和自动化编程：大模型如 Codex 可以根据自然语言描述生成代码，大大提升开发效率。开发者可以通过自然语言编写函数描述，然后模型自动生成代码。
    2. 错误检测和修复：大模型可以分析代码中的潜在错误，并给出修复建议，帮助程序员减少 Bug。
    3. 代码文档生成：大模型能够根据代码自动生成文档或注释，帮助提升代码的可读性和维护性。

3. 医疗与健康领域：大模型在医疗健康领域的应用潜力巨大，特别是在诊断、药物发现和患者护理等方面：

    1. 临床决策支持：大模型可以通过分析病历、医学文献和病患数据，辅助医生进行诊断和治疗决策，特别是在复杂病例中发挥作用。
    2. 药物发现：大模型能够通过对海量化学和生物数据的分析，发现新的药物组合或加速药物研发过程。
    3. 个性化医疗：通过分析病人的基因数据和病历，大模型可以提供个性化的治疗方案，优化健康管理和预防措施。

4. 金融与经济分析：在金融和经济领域，大模型有着深远的应用潜力：

    1. 自动化交易与投资建议：大模型能够通过分析金融数据、新闻和市场动态，进行自动化交易和投资建议的生成。它可以处理复杂的经济模型，并快速做出反应。
    2. 风险管理与欺诈检测：通过大数据和深度学习技术，大模型能够发现金融欺诈行为、预测市场波动并进行风险评估。
    3. 财务报告生成：大模型能够自动生成企业财务报告、市场分析和新闻稿，减少人力投入并提高效率。

5. 创意与艺术：大模型在艺术创作、设计和内容生成方面展现了很大的潜力：

    1. 自动化内容创作：大模型可以生成小说、诗歌、音乐歌词等创意内容，甚至可以根据特定风格生成符合要求的艺术作品。
    2. 图像生成：结合生成对抗网络（GAN）等技术，大模型可以用于生成高质量的图像、艺术品，甚至在广告和设计行业中应用。
    3. 游戏开发：大模型可以为游戏中的角色对话、剧情发展以及游戏环境生成提供创新支持，提升游戏的内容深度和复杂性。

6. 教育与个性化学习：在教育领域，大模型可以改变个性化学习的方式：

    1. 智能辅导：大模型能够通过自然语言理解为学生提供个性化的辅导和问题解答。它可以自动生成学习材料、解释复杂概念，并根据学生的学习进度进行调整。
    2. 语言学习：大模型可以成为语言学习者的虚拟老师，提供实时的对话练习和语言矫正，帮助学习者更有效地掌握语言技能。
    3. 教育内容生成：教师可以使用大模型自动生成试题、练习题、课程材料等，大幅减少课程准备的工作量。

7. 法律与合规：大模型在法律和合规领域也有潜力通过自动化提升效率：

    1. 法律文件生成：大模型可以自动撰写合同、法律意见书等法律文件，减少律师和法律顾问的工作量。
    2. 法规合规审查：大模型能够快速审查法律法规，帮助企业确保其运营符合相关法律规定，并自动更新合规要求。
    3. 案件分析：通过对以往判例的分析，大模型可以辅助律师进行案件分析和预判，提供策略支持。

8. 科学研究：大模型在科学研究中的应用也显示了潜力：

    1. 数据分析和模拟：大模型能够分析科学实验数据，帮助研究人员更好地理解复杂系统的行为，进行预测和模拟。
    2. 文献综述和撰写：大模型可以自动撰写文献综述、报告和科研论文，减少研究人员的写作负担。
    3. 跨学科合作：大模型能够将不同领域的知识整合，帮助科学家在跨学科领域发现新的研究方向。

9. 自动驾驶与机器人：大模型的推理和生成能力也为自动驾驶和机器人技术提供了新的可能：

    1. 自动驾驶：大模型可以通过多模态数据（如图像、激光雷达数据）训练和优化自动驾驶系统，提高其在复杂环境中的表现。
    2. 机器人对话与决策：大模型使机器人能够进行更自然的人机对话，并做出更智能的决策，提高机器人在服务行业的应用效果。

## 大模型和大语言模型

大语言模型（LLM）是大模型（LM）的一个子集，专门用于语言处理任务，而大模型的概念更广泛，涵盖各种数据类型和任务的处理能力。但架构和训练数据略有区别：

- 大模型（LM）的架构可以是多种多样的，具体取决于处理的数据类型和任务。如视觉领域的CNN、蛋白质折叠预测的AlphaFold等
- 大语言模型（LLM）通常基于Transformer等特定架构，并使用大量文本数据进行训练

[^1]: https://huggingface.co/blog/moe
